cardano-crypto-class/src/Cardano/Crypto/PinnedSizedBytes/TH.hs

```diff
@@ -0,0 +1,124 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TemplateHaskellQuotes #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Cardano.Crypto.PinnedSizedBytes.TH (
+  psbHex
+  ) where
+
+import Text.Read (readMaybe)
+import Data.Text (Text)
+import qualified Data.Text as Text
+import qualified Data.Kind as Haskell
+import Cardano.Crypto.PinnedSizedBytes.Internal (
+  PinnedSizedBytes (PSB)
+  )
+import Data.Primitive.ByteArray (
+  emptyByteArray, 
+  writeByteArray,
+  ByteArray,
+  unsafeFreezeByteArray,
+  newPinnedByteArray,
+  )
+import Control.Monad.ST (runST)
+import Language.Haskell.TH.Quote (
+  QuasiQuoter (
+    QuasiQuoter, 
+    quoteExp,
+    quotePat,
+    quoteType,
+    quoteDec
+    )
+  )
+import Language.Haskell.TH (
+  Q, 
+  Exp (AppTypeE, AppE, ConE, VarE), 
+  Type (LitT),
+  TyLit (NumTyLit),
+  )
+import Control.Monad.Trans.State.Strict (runStateT, StateT, get, modify)
+import Data.Word (Word8)
+import Data.Functor (($>))
+import Data.Foldable (traverse_)
+
+-- | Constructs (and verifies) a compile-time-known hex string as a
+-- 'PinnedSizedBytes', as well as computing its length at the type level.
+--
+-- The syntax accepted by this quasiquoter only works in an expression context.
+-- A valid input consists of the following:
+--
+-- * An optional @0x@;
+-- * An even number of /hexits/, which are either digits (@0@ through @9@) or
+-- letters (@a@ through @f@ or @A@ through @F@).
+--
+-- It is acceptable to mix upper-case and lower-case letters in the same input.
+-- We allow any amount of leading or trailing whitespace, but no internal
+-- whitespace.
+--
+-- This syntax produces a 'PinnedSizedBytes', of length equal to half the number
+-- of hexits, where the hexit pairs define a byte at that position. 
+--
+-- = Examples
+--
+-- * @['psbHex'| 0x |]@ produces a 'PinnedByteString' of length 0. This can also
+-- be written @['psbHex'|  |]@ or @['psbHex'|                |]@.
+-- * @['psbHex' | 0xafbc |]@ produces a 'PinnedByteString of length 2; the 
+-- first byte is @0xaf@, and the second is @0xbc@. This can also be written
+-- @['psbHex'| afbc |]@, @['psbHex'| AfBc |]@ or @['psbHex'|     Afbc    |]@.
+--
+-- = Note
+--
+-- To use this quasiquoter, @DataKinds@ and @TypeApplications@ need to be on.
+psbHex :: QuasiQuoter
+psbHex = QuasiQuoter {
+  quoteExp = mkHexLiteral,
+  quotePat = qqUseError "pattern",
+  quoteType = qqUseError "type",
+  quoteDec = qqUseError "declaration"
+  }
+
+-- Helpers
+
+mkHexLiteral :: String -> Q Exp
+mkHexLiteral input = do
+  let asStrippedText = Text.strip . Text.pack $ input
```

Comment:
This may be one of the rare cases where going through `Text` instead of keeping things as `String` isn't helpful. We're traversing the string exactly once, sequentially, and at each step, we just unpack a 2-character substring and feed it to `readMaybe`. Might as well do it all on a `String`. This isn't going to be performance critical anyway, since it runs at compile time, and getting rid of the `Text.{un}pack` bits would reduce visual clutter.

Reply:
The flip side is that stripping whitespace (leading and trailing) would get more difficult. I don't mind making this change if you insist, but I personally think it's easier to work with `Text` in this case.

Reply:
Fair enough.